[REEF-663] Make Http Endpoint available for client after submit

jwang98052 opened the pull request:
-This PR is to return HttpEndPoint to the client after Job is submitted
-At Java side, the http endpoint is written to JobSubmission folder in HDFS and Driver folder
-At REEFClient side, HttpClientHelper is added to pull the http endpoint from the file and return it in Submit()
-E2e test cases are updated to get the URI and verify on it.
-The new Submit API is noted as [Ustable] as we might change it in future
-The change is part of the change in tlc branch with some modifications

JIRA: [REEF-663](https://issues.apache.org/jira/browse/REEF-663)

This closes #

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Common/JavaClientLauncher.cs:
```diff
@@ -100,6 +100,7 @@ private string AssembleArguments(string javaClassName, params string[] parameter
             IList<string> arguments = new List<string>();
             arguments.Add("-cp");
             arguments.Add(GetClientClasspath());
+            arguments.Add("-Djava.util.logging.config.class=org.apache.reef.util.logging.Config");
```
This seems orthogonal to the goals of the PR. Comment?

jwang98052 replied:
I can remove for now.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Files/REEFFileNames.cs:
```diff
@@ -229,6 +230,11 @@ public string GetBridgeExeConfigPath()
+        /// <summary>
+        /// </summary>
+        /// <returns>File name that contains the dfs path for the DriverHttpEndpoint</returns>
+        public string DriverHttpEndpoint { get { return DRIVER_HTTP_ENDPOINT_FILE_NAME; } }
```
This needs to have an `[Unstable]` annotation, as it is likely to go away once we solve this for good.

jwang98052 replied:
Done

markusweimer commented:
I've done a pass. Two overarching comments:

  * `SubmitAsync` on a method that isn't `async` is odd. Also, it grabs the name such that we can't use it in future APIs. Maybe use `SubmitAndGetDriverUrl`?
  * I don't think we should change any of the examples or applications in the code base to use this API. It is temporary anyhow.

jwang98052 commented:
In the new Submit method in LocalClient, we do Launch Java with a separate  Task.Run(() =>. And then pull the result back. I have renamed it anyway as it is temporary. 
I have changed most places to use old Submit. Keep one test case to call the new  SubmitAndGetDriverUrl because I would like to have a test case for it.

jwang98052 commented:
The change has pushed in.

tcNickolas commented on lang/cs/Org.Apache.REEF.Client/Common/HttpClientHelper.cs:
```diff
@@ -0,0 +1,344 @@
+            return null;
+        }
+
+        private static bool ShouldRetry(HttpStatusCode httpStatusCode)
```
Is this function used? Looks like only the other version is used

tcNickolas commented on lang/cs/Org.Apache.REEF.Tests/Functional/ReefFunctionalTest.cs:
```diff
@@ -51,6 +51,7 @@ public class ReefFunctionalTest
 
         private const string Local = "local";
         private const string YARN = "yarn";
+        private const int SleppTime = 1000;
```
Typo: SleepTime

jwang98052 replied:
Done

tcNickolas commented on lang/cs/Org.Apache.REEF.Tests/Functional/ReefFunctionalTest.cs:
```diff
@@ -139,14 +140,35 @@ protected void ValidateSuccessForLocalRuntime(int numberOfEvaluatorsToClose, str
+            else
+            {
+                Console.WriteLine("Cannot read from log file");
+                Assert.IsNotNull(null);
```
Could be Assert.Fail() (which makes the intention a bit more evident)

jwang98052 replied:
Done

tcNickolas commented on lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java:
```diff
@@ -43,8 +44,8 @@
 import org.apache.reef.util.JARFileMaker;
-
 import javax.inject.Inject;
+import java.io.*;
```
If we're importing `io.*` we can remove subsequent imports of individual classes. But from our code it looks like we're trying not to import * but rather list individual classes we need.

jwang98052 replied:
Done

tcNickolas commented on lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java:
```diff
@@ -207,7 +206,65 @@ private void launch(final YarnSubmissionFromCS yarnSubmission) throws IOExceptio
+        LOG.log(Level.INFO, "Attempt " + i + " reading " + httpEndpointPath.toString());
+        if (fs.exists(httpEndpointPath)) {
+          FSDataInputStream input = fs.open(httpEndpointPath);
+          BufferedReader reader = new BufferedReader(new InputStreamReader(input));
```
In REEF-768 we want to eliminate high-priority warnings from Findbugs; this line will (likely) produce a new warning, `DM_DEFAULT_ENCODING: Reliance on default encoding` which complains about not setting encoding explicitly. Adding "UTF-8" as second parameter for `InputStreamReader` constructor should fix it.

jwang98052 replied:
Done

tcNickolas commented on lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java:
```diff
@@ -207,7 +206,65 @@ private void launch(final YarnSubmissionFromCS yarnSubmission) throws IOExceptio
+    }
+
+    final File driverHttpEndpointFile = new File(driverFolder, fileNames.getDriverHttpEndpoint());
+    BufferedWriter out = new BufferedWriter(new FileWriter(driverHttpEndpointFile));
```
Same Findbugs warning here; `new OutputStreamWriter(file, "UTF-8")` instead of `FileWriter` should fix it

jwang98052 replied:
OutputStreamWriter doesn't accept File as a parameter.

tcNickolas commented on lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/generic/JobDriver.java:
```diff
@@ -188,6 +198,17 @@ private void setupBridge(final ClrHandlersInitializer initializer) {
+      if (portNumber != null){
+        try {
+          final File outputFileName = new File(reefFileNames.getDriverHttpEndpoint());
+          BufferedWriter out = new BufferedWriter(new FileWriter(outputFileName));
```
Same warning

tcNickolas commented:
I've done a pass

jwang98052 commented:
Review comments addressed and changes pushed in.

markusweimer commented:
I'll test and merge.
